indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -377,6 +403,44 @@ static Granularity findGranularityFromSegments(List<DataSegment> segments)
     }
   }
 
+  /**
+   * Creates shard specs based on the given configurations. The return value is a map between intervals created
+   * based on the segment granularity and the shard specs to be created.
+   * Note that the shard specs to be created is a pair of {@link ShardSpecFactory} and number of segments per interval
+   * and filled only when {@link #isGuaranteedRollup} = true. Otherwise, the return value contains only the set of
+   * intervals generated based on the segment granularity.
+   */
+  protected static Map<Interval, Pair<ShardSpecFactory, Integer>> createShardSpecWithoutInputScan(
+      GranularitySpec granularitySpec,
+      IndexIOConfig ioConfig,
+      IndexTuningConfig tuningConfig,
+      PartitionsSpec nonNullPartitionsSpec
+  )
+  {
+    final Map<Interval, Pair<ShardSpecFactory, Integer>> allocateSpec = new HashMap<>();
+    final SortedSet<Interval> intervals = granularitySpec.bucketIntervals().get();
+
+    if (isGuaranteedRollup(ioConfig, tuningConfig)) {
+      // Overwrite mode, guaranteed rollup: shardSpecs must be known in advance.
+      assert nonNullPartitionsSpec instanceof HashedPartitionsSpec;
```

Comment:
is this because we need to run job like `DeterminePartitionsJob` to figure out partitions otherwise, which will add another phase and avoided right now?

Reply:
Hmm, the thing about `DeterminePartitionsJob` is correct, but this assertion is because the index task and the parallel index task currently only supports the hashed partitions spec. The range partitions spec will be supported as well and this assertion will be removed in the future.

Or, if you're asking about the comment, the index task already has a similar mode to determine partitions automatically and this method will not be called in that mode.

Reply:
I wasn't asking about the comment but was trying to understand what limits us technically from using dimension partition spec . it seems my guess was correct. it would be nice to add that in the comment.

Reply:
Added a comment.